concrete/src/User/UserInfo.php

```diff
@@ -330,104 +381,106 @@ public function getUserObject()
-            $session = Core::make('session');
-            if ($session->has('uID') && ($session->get('uID') == $this->getUserID())) {
-                $session->set('uName', $uName);
-                $session->set('uTimezone', $uTimezone);
-                $session->set('uDefaultLanguage', $uDefaultLanguage);
             }
-
-            // run any internal event we have for user update
-            $ue = new \Concrete\Core\User\Event\UserInfo($this);
-            Events::dispatch('on_user_update', $ue);
-
-            if ($testChange) {
-                $ue = new \Concrete\Core\User\Event\UserInfoWithPassword($this);
-                $ue->setUserPassword($data['uPassword']);
-                Events::dispatch('on_user_change_password', $ue);
+            if ($result === true && !empty($fields)) {
+                $this->connection->executeQuery(
+                    'update Users set  ' . implode(', ', $fields) . 'where uID = ? limit 1',
+                    array_merge($values, [$uID])
+                );
+                if ($emailChanged) {
+                    $this->connection->executeQuery('DELETE FROM UserValidationHashes WHERE uID = ?', [$uID]);
+                    $h = $this->application->make('helper/validation/identifier');
+                    $h->deleteKey('UserValidationHashes', 'uID', $uID);
+                }
+                // now we check to see if the user is updated his or her own logged in record
+                $session = $this->application->make('session');
+                if ($session->has('uID') && $uID === (int) $session->get('uID')) {
+                    if (isset($data['uName'])) {
+                        $session->set('uName', $data['uName']);
+                    }
+                    if (isset($data['uTimezone'])) {
+                        $session->set('uTimezone', $data['uTimezone']);
+                    }
+                    if (isset($data['uDefaultLanguage'])) {
+                        $session->set('uDefaultLanguage', $data['uDefaultLanguage']);
+                    }
+                    if ($passwordChangedOn !== null) {
+                        $session->set('uLastPasswordChange', $passwordChangedOn);
+                    }
+                }
+                // run any internal event we have for user update
+                $ue = new UserInfoEvent($this);
+                $this->application->make('director')->dispatch('on_user_update', $ue);
+                if ($passwordChangedOn !== null) {
+                    $ue = new UserInfoWithPasswordEvent($this);
+                    $ue->setUserPassword($data['uPassword']);
+                    $this->application->make('director')->dispatch('on_user_change_password', $ue);
+                }
             }
-
-            return $res;
         }
+
+        return $result;
     }
 
     /**
      * @param int[] $groupArray
      */
     public function updateGroups($groupArray)
```

Comment:
Avoid accessing undefined vars, throw `on_user_exit_group` event once per group (before we called that event twice (before and after deleting the records), optimize code, avoid potential security problems